In [ ]:
from pathlib import Path
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import ttest_ind
import requests
from io import StringIO

In [ ]:
# Set up data folder
DATA = Path('/content/drive/My Drive/data')

## Pathway mutation analysis

Use `KEGG REST API` to obtain all KEGG pathways and the genes.

In [ ]:
# Get pathway_id, gene_id table
KEGG_PATHWAY_LINK_BASE = "http://rest.kegg.jp/link/pathway/hsa"
pathway_link = requests.get(KEGG_PATHWAY_LINK_BASE)
kegg = pd.read_csv(StringIO(pathway_link.text), header=None, sep="\t")
kegg.columns = ["gene_id","pathway_id"]

# Get name of each pathway
KEGG_PATHWAY_LIST_BASE = "http://rest.kegg.jp/list/pathway/hsa"
pathway_list = requests.get(KEGG_PATHWAY_LIST_BASE)
kegg_name = pd.read_csv(StringIO(pathway_list.text), header=None,
                                    sep="\t")
kegg_name.columns = ["pathway_id","pathway_name"]

kegg = kegg.merge(kegg_name)
kegg.head()

gene_id     pathway_id                                       pathway_name
0  hsa:10327  path:hsa00010  Glycolysis / Gluconeogenesis - Homo sapiens (h...
1    hsa:124  path:hsa00010  Glycolysis / Gluconeogenesis - Homo sapiens (h...
2    hsa:125  path:hsa00010  Glycolysis / Gluconeogenesis - Homo sapiens (h...
3    hsa:126  path:hsa00010  Glycolysis / Gluconeogenesis - Homo sapiens (h...
4    hsa:127  path:hsa00010  Glycolysis / Gluconeogenesis - Homo sapiens (h...

In [ ]:
maf_file = DATA /"TCGA.BRCA.mutect.995c0111-d90b-4140-bee7-3845436c3b42.DR-10.0.somatic.maf.gz"

pd.set_option('display.max_columns', None)
maf = pd.read_csv(DATA / maf_file, sep="\t", comment='#', low_memory=False)
maf.head()

Hugo_Symbol  Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \
0       USP24           23358  WUGSC     GRCh38       chr1        55159655   
1      ERICH3          127254  WUGSC     GRCh38       chr1        74571494   
2      KIF26B           55083  WUGSC     GRCh38       chr1       245419680   
3       USP34            9736  WUGSC     GRCh38       chr2        61189055   
4      ANTXR1           84168  WUGSC     GRCh38       chr2        69245305   

   End_Position Strand Variant_Classification Variant_Type Reference_Allele  \
0      55159655      +      Missense_Mutation          SNP                T   
1      74571494      +      Missense_Mutation          SNP                C   
2     245419680      +                 Silent          SNP                G   
3      61189055      +                 Silent          SNP                G   
4      69245305      +                 Silent          SNP                G   

  Tumor_Seq_Allele1 Tumor_Seq_Allele2     dbSNP_RS  \
0                 T                 C  rs150880897   
1                 C                 T          NaN   
2                 G                 T          NaN   
3                 G                 C          NaN   
4                 G                 A  rs573467642   

                dbSNP_Val_Status          Tumor_Sample_Barcode  \
0  by1000G;byCluster;byFrequency  TCGA-D8-A1XY-01A-11D-A14K-09   
1                            NaN  TCGA-D8-A1XY-01A-11D-A14K-09   
2                            NaN  TCGA-D8-A1XY-01A-11D-A14K-09   
3                            NaN  TCGA-D8-A1XY-01A-11D-A14K-09   
4                        by1000G  TCGA-D8-A1XY-01A-11D-A14K-09   

    Matched_Norm_Sample_Barcode  Match_Norm_Seq_Allele1  \
0  TCGA-D8-A1XY-10A-01D-A14K-09                     NaN   
1  TCGA-D8-A1XY-10A-01D-A14K-09                     NaN   
2  TCGA-D8-A1XY-10A-01D-A14K-09                     NaN   
3  TCGA-D8-A1XY-10A-01D-A14K-09                     NaN   
4  TCGA-D8-A1XY-10A-01D-A14K-09                     NaN   

   Match_Norm_Seq_Allele2 Tumor_Validation_Allele1 Tumor_Validation_Allele2  \
0                     NaN                      NaN                      NaN   
1                     NaN                      NaN                      NaN   
2                     NaN                      NaN                      NaN   
3                     NaN                      NaN                      NaN   
4                     NaN                      NaN                      NaN   

   Match_Norm_Validation_Allele1  Match_Norm_Validation_Allele2  \
0                            NaN                            NaN   
1                            NaN                            NaN   
2                            NaN                            NaN   
3                            NaN                            NaN   
4                            NaN                            NaN   

   Verification_Status  Validation_Status Mutation_Status  Sequencing_Phase  \
0                  NaN                NaN         Somatic               NaN   
1                  NaN                NaN         Somatic               NaN   
2                  NaN                NaN         Somatic               NaN   
3                  NaN                NaN         Somatic               NaN   
4                  NaN                NaN         Somatic               NaN   

   Sequence_Source Validation_Method  Score  BAM_File            Sequencer  \
0              NaN               NaN    NaN       NaN  Illumina HiSeq 2000   
1              NaN               NaN    NaN       NaN  Illumina HiSeq 2000   
2              NaN               NaN    NaN       NaN  Illumina HiSeq 2000   
3              NaN               NaN    NaN       NaN  Illumina HiSeq 2000   
4              NaN               NaN    NaN       NaN  Illumina HiSeq 2000   

                      Tumor_Sample_UUID              Matched_Norm_Sample_UUID  \
0  edb6d161-8f50-4c11-8246-487c4ea9a55d  8dea96d9-5017-4872-a84e-33bfd2f37b7a 

In [ ]:
# print('Filter:', maf.FILTER.unique())
# print('Mutation_Status:', maf.Mutation_Status.unique())
# print('Variant_Classification:', maf.Variant_Classification.unique())
# print('IMPACT:', maf.IMPACT.unique())

In [ ]:
# make sure the mutations pass the "filter" and they are not modifiers
mut = maf[(maf.FILTER == 'PASS') & (maf.IMPACT != 'MODIFIER')].copy()

print("{} out of {} mutations passed the filtering criteria, {:.2f}%.".format(
    len(mut), len(maf), 100 * len(mut) / len(maf)))

# mut.head()

95711 out of 120988 mutations passed the filtering criteria, 79.11%.


The KEGG pathway table's gene IDs has `hsa:` prefix to them.

In [ ]:
# add hsa to the id in maf table so it can match with kegg

mut_hsa = mut.copy()
mut_hsa.Entrez_Gene_Id = mut_hsa.apply(lambda x: 'hsa:'+ str(x.Entrez_Gene_Id), axis = 1)
len(mut_hsa)

95711

In [ ]:
mut_hsa.head()

Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \
1      ERICH3     hsa:127254  WUGSC     GRCh38       chr1        74571494   
2      KIF26B      hsa:55083  WUGSC     GRCh38       chr1       245419680   
3       USP34       hsa:9736  WUGSC     GRCh38       chr2        61189055   
4      ANTXR1      hsa:84168  WUGSC     GRCh38       chr2        69245305   
5       SCN9A       hsa:6335  WUGSC     GRCh38       chr2       166199365   

   End_Position Strand Variant_Classification Variant_Type Reference_Allele  \
1      74571494      +      Missense_Mutation          SNP                C   
2     245419680      +                 Silent          SNP                G   
3      61189055      +                 Silent          SNP                G   
4      69245305      +                 Silent          SNP                G   
5     166199365      +                 Silent          SNP                G   

  Tumor_Seq_Allele1 Tumor_Seq_Allele2     dbSNP_RS dbSNP_Val_Status  \
1                 C                 T          NaN              NaN   
2                 G                 T          NaN              NaN   
3                 G                 C          NaN              NaN   
4                 G                 A  rs573467642          by1000G   
5                 G                 A          NaN              NaN   

           Tumor_Sample_Barcode   Matched_Norm_Sample_Barcode  \
1  TCGA-D8-A1XY-01A-11D-A14K-09  TCGA-D8-A1XY-10A-01D-A14K-09   
2  TCGA-D8-A1XY-01A-11D-A14K-09  TCGA-D8-A1XY-10A-01D-A14K-09   
3  TCGA-D8-A1XY-01A-11D-A14K-09  TCGA-D8-A1XY-10A-01D-A14K-09   
4  TCGA-D8-A1XY-01A-11D-A14K-09  TCGA-D8-A1XY-10A-01D-A14K-09   
5  TCGA-D8-A1XY-01A-11D-A14K-09  TCGA-D8-A1XY-10A-01D-A14K-09   

   Match_Norm_Seq_Allele1  Match_Norm_Seq_Allele2 Tumor_Validation_Allele1  \
1                     NaN                     NaN                      NaN   
2                     NaN                     NaN                      NaN   
3                     NaN                     NaN                      NaN   
4                     NaN                     NaN                      NaN   
5                     NaN                     NaN                      NaN   

  Tumor_Validation_Allele2  Match_Norm_Validation_Allele1  \
1                      NaN                            NaN   
2                      NaN                            NaN   
3                      NaN                            NaN   
4                      NaN                            NaN   
5                      NaN                            NaN   

   Match_Norm_Validation_Allele2  Verification_Status  Validation_Status  \
1                            NaN                  NaN                NaN   
2                            NaN                  NaN                NaN   
3                            NaN                  NaN                NaN   
4                            NaN                  NaN                NaN   
5                            NaN                  NaN                NaN   

  Mutation_Status  Sequencing_Phase  Sequence_Source Validation_Method  Score  \
1         Somatic               NaN              NaN               NaN    NaN   
2         Somatic               NaN              NaN               NaN    NaN   
3         Somatic               NaN              NaN               NaN    NaN   
4         Somatic               NaN              NaN               NaN    NaN   
5         Somatic               NaN              NaN               NaN    NaN   

   BAM_File            Sequencer                     Tumor_Sample_UUID  \
1       NaN  Illumina HiSeq 2000  edb6d161-8f50-4c11-8246-487c4ea9a55d   
2       NaN  Illumina HiSeq 2000  edb6d161-8f50-4c11-8246-487c4ea9a55d   
3       NaN  Illumina HiSeq 2000  edb6d161-8f50-4c11-8246-487c4ea9a55d   
4       NaN  Illumina HiSeq 2000  edb6d161-8f50-4c11-8246-487c4ea9a55d   
5       NaN  Illumina HiSeq 2000  edb6d161-8f50-4c11-8246-487c4ea9a55d   

               Matche

In [ ]:
path_id = []
num_mut = []
pathway_name = []

for name, group in kegg.groupby('pathway_id'):
  path_id.append(name)
  num_mut.append(sum(mut_hsa.Entrez_Gene_Id.isin(group.gene_id)))
  pathway_name.append(group.pathway_name)
  # print(name)

In [ ]:
# # this cell takes too long, so use the one above

# # for n, g in kegg.groupby('pathway_id'):
# #     print(n, g)

# path_id = []
# num_mut = []

# for name, group in kegg.groupby('pathway_id'):
#   path_id.append(name)
#   num_mut.append(sum(x in list(group.gene_id) for x in mut_hsa.Entrez_Gene_Id))
# #     print(name, list(group.gene_id))
# #     print(name, sum(x in list(group.gene_id) for x in mut_hsa.Entrez_Gene_Id))

In [ ]:
path_count = np.c_[path_id, num_mut, pathway_name]
path_count = pd.DataFrame(path_count, columns = ['pathway_id', 'num_samples', 'pathway_name'])
print(path_count.shape)  
path_count.head()

(336, 3)


pathway_id num_samples  \
0  path:hsa00010         241   
1  path:hsa00020         131   
2  path:hsa00030         111   
3  path:hsa00040         125   
4  path:hsa00051         133   

                                        pathway_name  
0  0     Glycolysis / Gluconeogenesis - Homo sapi...  
1  68    Citrate cycle (TCA cycle) - Homo sapiens...  
2  98     Pentose phosphate pathway - Homo sapien...  
3  128    Pentose and glucuronate interconversion...  
4  162    Fructose and mannose metabolism - Homo ...

The cell below gives the number of breast cancer samples that have non-MODIFIER mutations in each KEGG pathway. They are ranked.

In [ ]:
path_count.num_samples = path_count.num_samples.astype(int)
path_count.sort_values('num_samples', ascending=False)

pathway_id  num_samples  \
87   path:hsa01100         6593   
297  path:hsa05200         4055   
294  path:hsa05168         3239   
291  path:hsa05165         3159   
147  path:hsa04151         3137   
115  path:hsa04010         2246   
169  path:hsa04510         2237   
301  path:hsa05205         2164   
292  path:hsa05166         2101   
121  path:hsa04024         1949   
302  path:hsa05206         1918   
289  path:hsa05163         1895   
118  path:hsa04015         1886   
287  path:hsa05161         1808   
215  path:hsa04810         1766   
225  path:hsa04919         1754   
318  path:hsa05225         1695   
299  path:hsa05203         1695   
319  path:hsa05226         1688   
211  path:hsa04740         1686   
293  path:hsa05167         1661   
317  path:hsa05224         1656   
119  path:hsa04020         1651   
163  path:hsa04360         1647   
295  path:hsa05169         1622   
155  path:hsa04218         1613   
117  path:hsa04014         1581   
315  path:hsa05222         1522   
149  path:hsa04210         1506   
93   path:hsa01522         1493   
..             ...          ...   
29   path:hsa00360           72   
76   path:hsa00790           71   
48   path:hsa00531           68   
11   path:hsa00100           68   
35   path:hsa00450           67   
107  path:hsa03060           62   
8    path:hsa00062           62   
328  path:hsa05330           58   
89   path:hsa01210           57   
33   path:hsa00430           50   
62   path:hsa00603           45   
69   path:hsa00730           44   
63   path:hsa00604           43   
323  path:hsa05310           42   
50   path:hsa00533           42   
10   path:hsa00072           42   
70   path:hsa00740           34   
13   path:hsa00130           29   
81   path:hsa00920           28   
47   path:hsa00524           27   
71   path:hsa00750           23   
34   path:hsa00440           22   
18   path:hsa00232           19   
74   path:hsa00780           15   
36   path:hsa00471           15   
31   path:hsa00400           15   
24   path:hsa00290           13   
75   path:hsa00785           11   
136  path:hsa04122           11   
37   path:hsa00472            5   

                                          pathway_name  
87   2966    Metabolic pathways - Homo sapiens (hum...  
297  27318    Pathways in cancer - Homo sapiens (hu...  
294  26414    Herpes simplex virus 1 infection - Ho...  
291  25679    Human papillomavirus infection - Homo...  
147  10932    PI3K-Akt signaling pathway - Homo sap...  
115  6241    MAPK signaling pathway - Homo sapiens ...  
169  13499    Focal adhesion - Homo sapiens (human)...  
301  28318    Proteoglycans in cancer - Homo sapien...  
292  26009    Human T-cell leukemia virus 1 infecti...  
121  7423    cAMP signaling pathway - Homo sapiens ...  
302  28522    MicroRNAs in cancer - Homo sapiens (h...  
289  25284    Human cytomegalovirus infection - Hom...  
118  6853    Rap1 signaling pathway - Homo sapiens ...  
287  24984    Hepatitis B - Homo sapiens (human)
24...  
215  18582    Regulation of actin cytoskeleton - Ho...  
225  19642    Thyroid hormone signaling pathway - H...  
318  29954    Hepatocellular carcinoma - Homo sapie...  
299  28035    Viral carcinogenesis - Homo sapiens (...  
319  30122    Gastric cancer - Homo sapiens (human)...  
211  17923    Olfactory transduction - Homo sapiens...  
293  26228    Kaposi sarcoma-associated herpesvirus...  
317  29807    Breast cancer - Homo sapiens (human)
...  
119  7063    Calcium signaling pathway - Homo sapie...  
163  12811    Axon guidance - Homo sapiens (human)
...  
295  26905    Epstein-Barr virus infection - Homo s...  
155  11927    Cellular senescence - Homo sapiens (h...  
117  6621    Ras signaling pathway - Homo sapiens (...  
315  29649    Small cell lung cancer - Homo sapiens...  
149  11406    Apoptosis - Homo sapiens (human)
1140...  
93   4799    Endocrine resistance - Homo sapiens (h...  
..                                                 ...  
29   1152    

The following cells are trying to calculate: Number of patients with mutated pathway / Total number of patients.

In [ ]:
# Get the gene IDs under the top pathway from kegg
kegg_top = kegg[kegg.pathway_id == 'path:hsa01100']
kegg_top.head()

gene_id     pathway_id                               pathway_name
2966         hsa:10  path:hsa01100  Metabolic pathways - Homo sapiens (human)
2967        hsa:100  path:hsa01100  Metabolic pathways - Homo sapiens (human)
2968      hsa:10005  path:hsa01100  Metabolic pathways - Homo sapiens (human)
2969      hsa:10007  path:hsa01100  Metabolic pathways - Homo sapiens (human)
2970  hsa:100137049  path:hsa01100  Metabolic pathways - Homo sapiens (human)

In [ ]:
# Find the gene IDs from cancer samples that match the gene IDs under the top pathway
mut_top = mut_hsa[mut_hsa.Entrez_Gene_Id.isin(kegg_top.gene_id)]
mut_top.head()

Hugo_Symbol Entrez_Gene_Id Center NCBI_Build Chromosome  Start_Position  \
11      PIK3CA       hsa:5290  WUGSC     GRCh38       chr3       179218303   
12       MCCC1      hsa:56922  WUGSC     GRCh38       chr3       183092485   
14        UGT8       hsa:7368  WUGSC     GRCh38       chr4       114665726   
21        GUSB       hsa:2990  WUGSC     GRCh38       chr7        65974700   
28       EHMT1      hsa:79813  WUGSC     GRCh38       chr9       137777960   

    End_Position Strand Variant_Classification Variant_Type Reference_Allele  \
11     179218303      +      Missense_Mutation          SNP                G   
12     183092485      +      Missense_Mutation          SNP                C   
14     114665727      +        Frame_Shift_Del          DEL               AT   
21      65974700      +      Missense_Mutation          SNP                C   
28     137777960      +                 Silent          SNP                G   

   Tumor_Seq_Allele1 Tumor_Seq_Allele2     dbSNP_RS dbSNP_Val_Status  \
11                 G                 A  rs104886003        byCluster   
12                 C                 T  rs569042803          by1000G   
14                AT                 -          NaN              NaN   
21                 C                 T          NaN              NaN   
28                 G                 A  rs756140987      byFrequency   

            Tumor_Sample_Barcode   Matched_Norm_Sample_Barcode  \
11  TCGA-D8-A1XY-01A-11D-A14K-09  TCGA-D8-A1XY-10A-01D-A14K-09   
12  TCGA-D8-A1XY-01A-11D-A14K-09  TCGA-D8-A1XY-10A-01D-A14K-09   
14  TCGA-D8-A1XY-01A-11D-A14K-09  TCGA-D8-A1XY-10A-01D-A14K-09   
21  TCGA-D8-A1XY-01A-11D-A14K-09  TCGA-D8-A1XY-10A-01D-A14K-09   
28  TCGA-D8-A1XY-01A-11D-A14K-09  TCGA-D8-A1XY-10A-01D-A14K-09   

    Match_Norm_Seq_Allele1  Match_Norm_Seq_Allele2 Tumor_Validation_Allele1  \
11                     NaN                     NaN                      NaN   
12                     NaN                     NaN                      NaN   
14                     NaN                     NaN                      NaN   
21                     NaN                     NaN                      NaN   
28                     NaN                     NaN                      NaN   

   Tumor_Validation_Allele2  Match_Norm_Validation_Allele1  \
11                      NaN                            NaN   
12                      NaN                            NaN   
14                      NaN                            NaN   
21                      NaN                            NaN   
28                      NaN                            NaN   

    Match_Norm_Validation_Allele2  Verification_Status  Validation_Status  \
11                            NaN                  NaN                NaN   
12                            NaN                  NaN                NaN   
14                            NaN                  NaN                NaN   
21                            NaN                  NaN                NaN   
28                            NaN                  NaN                NaN   

   Mutation_Status  Sequencing_Phase  Sequence_Source Validation_Method  \
11         Somatic               NaN              NaN               NaN   
12         Somatic               NaN              NaN               NaN   
14         Somatic               NaN              NaN               NaN   
21         Somatic               NaN              NaN               NaN   
28         Somatic               NaN              NaN               NaN   

    Score  BAM_File            Sequencer  \
11    NaN       NaN  Illumina HiSeq 2000   
12    NaN       NaN  Illumina HiSeq 2000   
14    NaN       NaN  Illumina HiSeq 2000   
21    NaN       NaN  Illumina HiSeq 2000   
28    NaN       NaN  Illumina HiSeq 2000   

                       Tumor_Sample_UUID  \
11  edb6d161-8f50-4c11-8246-487c4ea9a55d   
12  edb6d161-8f50-4c11-8246-487c4ea9a55d   
14  edb6d161-8f50-4c11-8246-487c4ea9a55d   
21  edb6d161-8f50-4c

In [ ]:
num_top = mut_top.Tumor_Sample_Barcode.nunique()
num_top
# Number of patients with mutated pathway

924

In [ ]:
num_total = maf.Tumor_Sample_Barcode.nunique()
num_total
# Total number of patients

986

In [ ]:
mut_freq = num_top/num_total
mut_freq
# the mutation frequency of the pathway (Number of patients with mutated pathway / Total number of patients)

0.9371196754563894

The most mutated pathway in the TCGA breast cancer data set is hsa01100 (2966 Metabolic pathways - Homo sapiens). The mutation frequency of this pathway is 0.937 = 93.7%.